hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/index/bucket/HoodieSparkConsistentBucketIndex.java

```diff
@@ -275,4 +280,49 @@ public Option<HoodieRecordLocation> getRecordLocation(HoodieKey key) {
       throw new HoodieIndexException("Failed to getBucket as hashing node has no file group");
     }
   }
+
+  /**
+   * Update default metadata file(00000000000000.hashing_meta) with the latest committed metadata file so that default file will be in sync
+   * with latest commit.
+   *
+   * @param table
+   */
+  public void updateArchivalDependentIndexMetadata(HoodieTable table,List<HoodieInstant> hoodieArchivalInstants) {
+    Map<String, Boolean> partitionVisiteddMap = new HashMap<>();
+    // Update metadata for replace commit which are going to get archived.
+    Stream<HoodieInstant> hoodieListOfReplacedInstants = hoodieArchivalInstants.stream().filter(instane -> instane.getAction().equals(REPLACE_COMMIT_ACTION));
+    hoodieListOfReplacedInstants.forEach(instant -> {
+      Option<Pair<HoodieInstant, HoodieClusteringPlan>> instantPlanPair =
+          ClusteringUtils.getClusteringPlan(table.getMetaClient(), HoodieTimeline.getReplaceCommitRequestedInstant(instant.getTimestamp()));
+      if (instantPlanPair.isPresent()) {
+        HoodieClusteringPlan plan = instantPlanPair.get().getRight();
+        List<Map<String, String>> partitionMapList = plan.getInputGroups().stream().map(HoodieClusteringGroup::getExtraMetadata).collect(Collectors.toList());
+        partitionMapList.stream().forEach(partitionMap -> {
+          String partition = partitionMap.get(SparkConsistentBucketClusteringPlanStrategy.METADATA_PARTITION_KEY);
+          if (!partitionVisiteddMap.containsKey(partition)) {
+            Option<HoodieConsistentHashingMetadata> hoodieConsistentHashingMetadataOption = loadMetadata(table, partition);
+            if (hoodieConsistentHashingMetadataOption.isPresent()) {
```

Comment:
Still confused why we defer the metadata fix until archiving, shouldn't we update the consistent hashing metadata right after we finishing the clustering commit?

Reply:
@danny0405 , There are consistency issues we might face if we try updating metadata after finishing clustering commit.
What if underlying file system is down and updateMetadata fails to sync metadata, then there is no mechanism to bring it in sync with latest committed metadata, and archival will remove replace commit eventually and it will end up in an inconsistent state.
On the other hand in archival process , it will be eventually in sync with committed metadata before replace commit getting archived.
I think consistent hashing metadata has strong dependency on archival process, As it is dependent on active timeline replaced commit to load metadata.

Reply:
let me give few examples which can cause inconsistency
Example 1
lets say we are updating metadat before clustering commit
1 - written the data files on storage
2 - update metatda
3 - commited clustering operation on hudi timeline by creating replace commit

now let's say **2nd operation is done but 3rd operation fails** in this case **metadata got synced but clustering is failed. all the subsequent write operation will read from failed clustering metadata operation.** 

Example 2
lets say we are updating metadat after clustering commit
1 -  written the data files on storage
2 - commited clustering operation on hudi timeline by creating replace commit
3 - update metadata

now let's say 3rd operation of updating metadata fails , then latest metadata commited file will not be in sync default metadata file (0000000*.meta),and there is no scheduled mechanism to bring it in sync state. hence once archival archives latest metadata file related replace commit from timeline, all the writer will start reading from default metadata file (0000000*.meta) and this will cause data duplication.

Reply:
> What if underlying file system is down and updateMetadata fails to sync metadata, then there is no mechanism to bring it in sync with latest committed metadat

Just put the update into the same transaction of the clustering operation for consistency hash indexing.
That is, update the metadata first, then transite the state from inflight to complete for this replace commit.

Reply:
@danny0405 , Understood, one question lets say transaction fails at transitioning  the state from inflight to complete replace commit, does **update to metadata file(default file 00000*.meta)  operation get reverted**?

Reply:
yes, but i guess we do not have very good manner to put them into one atomic operation, we can do some adjust like when job recovering, fix the metadata if there is in-consistency. but in general, fail for file transition is a very rare case I think.

Reply:
Okay, will add the trigger in clustering operation itself. let me see how we can do that

Reply:
@danny0405 @SteNicholas modified the approach.
Creating commit marker file for completed clustering operation post commit. 
Reader will use commit marker as indicator to get latest metadata for consistent hashing.
If commit marker fails post commit, there is recovery job which will decide if its a valid latest metadata and create the marker file.

Please review.